### PR TITLE
selectively add setuptools_git as a setup_requires

### PR DESCRIPTION
**Pull request by tisdall**
@mcdonc seems to want to avoid a blanket addition of `setuptools_git` as a `setup_requires` so I'm suggesting this as a compromise...  This essentially adds it only if there's a `.git` directory.  This avoids the dependency if the person is installing via a pypi package.

The only issue with this is if a person checks out the repo and does a "setup.py install" on it, it will temporarily install `setuptools_git` even though it's not needed.

Issues this solves: #146 Pylons/pyramid#121 Pylons/pyramid#851 Pylons/substanced#32 repoze/repoze.tm2#4  #174 Pylons/substanced#177  Pylons/pyramid#1213 Pylons/pyramid_chameleon#7

As you can see, this keeps coming up over and over and over again.

### CODE REVIEW

**Comment by tisdall** [comment]
also Pylons/colander#132

**Comment by mmerickel** [comment]
The issue is that `setup_requires` is totally half-baked and doesn't work as advertised. There's a chicken and egg problem where `setup_requires` must be installed prior to running `setup.py`.

@dholth has tried to solve this with https://bitbucket.org/dholth/setup-requires.

I'll grant you that for this situation it would work fine because the actual `setup.py` does not require any external packages to run, but it's still a reason to avoid using `setup_requires` as it stands.

**Comment by tisdall** [comment]
@mmerickel I did say it was a _compromise_.  :smile: 

Basically I wasted a good chunk of time trying to do something I thought was simple like this:
```
from setuptools import setup

setup(
    install_requires=['deform'],
    dependency_links = [
        "git+https://github.com/Pylons/deform.git#egg=deform",
    ]
)
```

... and it not being clear why there where missing files.  I'm also not the first and I also won't be the last.  I'd like to put something in place to avoid other people wasting their time too.

I guess this would be the bandaid patch until someone figures out how to automatically maintain a MANIFEST.in file, no?

**Comment by mmerickel** [comment]
`dependency_links` are all but deprecated at this point and are bad form to use in a `setup.py` file regardless. Note that `pip` now requires an extra `--process-dependency-links` flag to use them.

The issue is that the `setup.py` should say how to build your think, but not where that stuff comes from. That is the job of something else (like `requirements.txt`). It is the same argument as not putting version pins in your `setup.py`, which is hopefully obvious.

You could put something like `-e git+https://github.com/Pylons/deform.git#egg=deform` in a requirements.txt and it should be installed exactly as you'd like, and your project will find it when it looks for the `install_requires=['deform']` happily.

**Comment by tisdall** [comment]
Hmm..  I tried what you suggested.  It seems that instead of some sort of installation in lib/python*/site-packages/ I have an egg-link to src/deform/ .  There's a `pip-delete-this-directory.txt` file in /src that says:

> This file is placed here by pip to indicate the source was put
> here by pip.
> 
> Once this package is successfully installed this source code will be
> deleted (unless you remove this file).

It seems pip installed via "setup.py develop"...

I tried changing it to just adding `git+https://github.com/Pylons/deform.git#egg=deform` so it properly does an install, but then I'm back to having an install missing many files.

**Comment by mmerickel** [comment]
Yes, you're trying to use an unreleased version of deform so you should install it with develop in editable mode. The only supported method of installing pylons project packages is from a distribution uploaded to PyPI unless you want to go through the effort of building it yourself. This entire discussion should be in Pylons/pyramid#121 since nothing will change here until it changes there.

**Comment by tisdall** [comment]
I just finished reading https://caremad.io/2013/07/setup-vs-requirement/ so I understand what you're saying about `dependency_links` now.  However the issue is still the same when you have pip fetch and install from the repository.

**Comment by rbu** [comment]
```dependency_links``` being deprecated is an orthogonal issue. You would have the exact same issue that triggered this PR and issue #146 when you work with pip's best practice of requirement.txt files.

**Comment by tisdall** [comment]
@rbu I think this is less of an issue with pip/requirements.txt , though.  It seems like the normal way of included git projects is through `-e git+https://github.com/...` which will include it using a `setup.py develop` which _will_ include everything needed.  However, it's still an issue if you include it in the requirement.txt as `git+https://github.com/...` (without the `-e` option) which installs it as `setup.py install`.

It's also "broken" if you call `pip install git+https://github.com/Pylons/deform.git`.  That checks out the repo in a tmp directory and calls `setup.py install` on it.

Basically, I think it's a bad practice to leave this pothole for people to stumble in when there's an easy (though somewhat ugly) patch for it.

**Comment by rbu** [comment]
I fully agree. It took some time for me to find out why sdist does not work as I am used with most of the other projects that I build myself. This was in 2012 (when I pushed https://github.com/Pylons/deform/commit/b83208682fe8134b65b8409674b3f24d7ef73c62). 

I would say that your list of issues in https://github.com/Pylons/deform/pull/257#issue-54496316 plus my list of issues in https://github.com/Pylons/deform/issues/146#issuecomment-62554722 plus all the references in https://github.com/Pylons/pyramid/issues/121 should be a good enough argument to merge.

**Comment by tisdall** [comment]
The `setup_requires` was added by you then removed only because it stopped working.  Then when it was re-added @mcdonc argued that it shouldn't be there because it'd be an unnecessary requirement for all PyPi users (I agree, it'd be nice to avoid that).  So, this seems to be the most ideal solution...  `setuptools-git` is completely useless without the `.git` meta data anyway, so we should only be trying to use it if that directory is there.

**Comment by tisdall** [comment]
closing and moving to #273
